src/graph/GraphRunCtx.hpp

```diff
@@ -0,0 +1,103 @@
+// Copyright 2022 Robotec.AI
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <list>
+#include <set>
+#include <vector>
+#include <thread>
+
+#include <CudaStream.hpp>
+#include <graph/Node.hpp>
+#include <graph/NodesCore.hpp>
+
+/**
+ * Structure storing context for running a graph.
+ * This is a 'volatile' struct - changing graph's structure destroys this context.
+ */
+struct GraphRunCtx
+{
+	static std::shared_ptr<GraphRunCtx> createAndAttach(std::shared_ptr<Node> node);
+
+	/**
+	 * Executes graph bound with this GraphRunCtx.
+	 */
+	void executeAsync();
+
+	/**
+	* Resets graphRunCtx for all nodes currently bound with this GraphRunCtx, consequently destroying this GraphRunCtx.
+	*/
+	void detachAndDestroy();
+
+	/**
+	 * Waits until this GraphRunCtx
+	 * - finishes execution
+	 * - joins its thread
+	 * - synchronizes graph stream (all pending GPU operations)
+	 */
+	void synchronize();
+
+	/**
+	 * Ensures that no GraphRunCtx is running.
+	 */
+	static void synchronizeAll();
+
+	/**
+	 * Waits until given node finishes its CPU execution.
+	 * The node may still have pending GPU operations.
+	 */
+	void synchronizeNodeCPU(Node::ConstPtr nodeToSynchronize);
+
+	bool isThisThreadGraphThread() const {return maybeThread.has_value() && maybeThread->get_id() == std::this_thread::get_id(); }
+
+	CudaStream::Ptr getStream() const { return stream; }
+	const std::set<std::shared_ptr<Node>>& getNodes() const { return nodes; }
+
+	virtual ~GraphRunCtx();
+private:
```

Comment:
Generally, I am against this, because it assumes/allows splitting class members of the same kind. Above, you have a constructor, a static method, some other method, few fields, then a static field, then nested class, and later there are again some fields. You are on point - this may improve overall understanding of class purposes, but IMO it lowers the readability. It may be be harder e.g. to see class members initialization order. Or to check, if some field is static or not.

This may be easier to handle via composition, if it is a matter of context (not only fields / methods split) - separate helper class responsible for some communication. On the other hand, fields / methods split is rather natural for most developers.

Reply:
Ok, removed 2 redundant `private`, with only a comment left.